text/profiles/0239-profiles-data-model.md

```diff
@@ -0,0 +1,1604 @@
+
+* by embedding a message into another message (standard protobuf way)
+* by referencing a message by index (similar to how it's done in pprof)
+
+Profiling signal is different from most other ones in that we use the referencing technique a lot to represent relationships between messages where there is a lot of duplication happening. This allows to reduce the size of the resulting protobuf payload and the number of objects that need to be allocated to parse such payload.
+
+This example illustrates the conceptual difference between the two approaches. Note that this example is simplified for clarity and provided in YAML format for legibility:
+
+```yaml
+# denormalized
+samples:
+- stacktrace:
+  - foo
+  - bar
+  value: 100
+  attribute_set:
+    endpoint: "/v1/users"
+- stacktrace:
+  - foo
+  - bar
+  - baz
+  value: 200
+  attribute_set:
+    endpoint: "/v1/users"
+
+# normalized
+attribute_sets:
+- endpoint: "/v1/users"
+samples:
+- stacktrace:
+  - foo
+  - bar
+  value: 100
+  attribute_set_index: 0
+- stacktrace:
+  - foo
+  - bar
+  - baz
+  value: 200
+  attribute_set_index: 0
+
+```
+
+Explanation: because multiple samples have the same attributes, we can store them in a separate table and reference them by index. This reduces the size of the resulting protobuf payload and the number of objects that need to be allocated to parse such payload.
+
+Benchmarking shows that this approach is significantly more efficient in terms of CPU utilization, memory consumption and size of the resulting protobuf payload. See [Prior art and alternatives](#prior-art-and-alternatives) for more details.
+
+#### Relationship Between Samples and Locations
+
+Relationship between Samples and Locations is using the referencing technique described above. However, there's an additional optimization technique used to reduce the size of the resulting protobuf payload and the number of objects that need to be allocated to parse such payload. The technique is based on the fact that many samples share the same locations.
+
+Considering the following example presented in a folded format:
+
+```
+foo;bar;baz 100
+abc;def 200
+foo;bar 300
+```
+
+It represents 3 samples:
+
+* one for stacktrace `foo;bar;baz` with value `100`.
+* one for stacktrace `abc;def` with value `200`.
+* one for stacktrace `foo;bar` with value `300`. Note that 2 of the locations are shared with the first sample.
+
+By storing locations in a separate table and referring to start_index+length in that table, we can reduce the size of the resulting protobuf payload and the number of objects that need to be allocated to parse such payload. With this approach we can also take advantage of the fact that many samples share the same locations. Below is a representation of the resulting protobuf payload (in YAML format for legibility):
+
+```yaml
+sample:
+  - locations_start_index: 0
+    locations_length: 3
+    value:
+      - 100
+  - locations_start_index: 2
```

Comment:
For `abc;def` the `locations_start_index` should be `4` 
```suggestion
  - locations_start_index: 4
```

Reply:
Would it be possible to address this on a follow-up PR?

Reply:
As this is just a typo, I think it does not change the purpose or intention of this PR and should be good to be fixed before it get merged.

Reply:
Let's address this in a follow up PR.